src/plugins/vis_augmenter/public/vega/helpers.ts

```diff
@@ -0,0 +1,324 @@
+              const higherValDiff = Math.abs(timestamp - higherVal);
+              rowIndexToInsert = smallerValDiff <= higherValDiff ? rowIndex : rowIndex + 1;
+            }
+
+            // timestamp is on the right bounds of the chart
+            else if (rowIndex + 1 === augmentedTable.rows.length - 1) {
+              rowIndexToInsert = rowIndex + 1;
+              // timestamp is still too small; traverse to next bucket
+            } else {
+              rowIndex += 1;
+              continue;
+            }
+
+            // inserting the value. increment if the mapping/property already exists
+            augmentedTable.rows[rowIndexToInsert][visLayerColumnId] =
+              (get(augmentedTable.rows[rowIndexToInsert], visLayerColumnId, 0) as number) + 1;
+            break;
+          }
+        });
+      }
+      return true;
+    });
+  }
+  return augmentedTable;
+};
+
+/**
+ * Updating the vega lite spec to include layers and marks related to
+ * PointInTimeEventsVisLayers. It is assumed the datatable has already been
+ * augmented with columns and row data containing the vis layers.
+ */
+export const addPointInTimeEventsLayersToSpec = (
+  datatable: OpenSearchDashboardsDatatable,
+  dimensions: any,
+  spec: object
+): object => {
+  const newSpec = cloneDeep(spec) as any;
+  newSpec.config = enableEventsInConfig(newSpec.config);
+
+  const xAxisId = getXAxisId(dimensions, datatable.columns);
+  const xAxisTitle = dimensions.x.label.replaceAll('"', '');
+  const bucketStartTime = convertToDateTimeObj(datatable.rows[0][xAxisId] as number);
+  const bucketEndTime = convertToDateTimeObj(
+    datatable.rows[datatable.rows.length - 1][xAxisId] as number
+  );
+  const visLayerColumns = datatable.columns.filter((column: OpenSearchDashboardsDatatableColumn) =>
+    isVisLayerColumn(column)
+  );
+  const visLayerColumnIds = visLayerColumns.map((column) => column.id);
+
+  // Hide x axes text on existing chart so they are only visible on the event chart
+  newSpec.layer.forEach((dataSeries: any) => {
+    if (get(dataSeries, 'encoding.x.axis', null) !== null) {
+      dataSeries.encoding.x.axis = {
+        ...dataSeries.encoding.x.axis,
+        labels: false,
+        title: null,
+      };
+    }
+  });
+
+  // Add a rule to the existing layer for showing lines on the chart if a dot is hovered on
+  newSpec.layer.push({
+    mark: {
+      type: 'rule',
+      color: EVENT_COLOR,
+      opacity: 1,
+    },
+    transform: [{ filter: generateVisLayerFilterString(visLayerColumnIds) }],
+    encoding: {
+      x: {
+        field: xAxisId,
+        type: 'temporal',
+      },
+      opacity: {
+        value: 0,
+        condition: { empty: false, param: HOVER_PARAM, value: 1 },
+      },
+    },
+  });
+
+  // Nesting layer into a vconcat field so we can append event chart.
+  newSpec.vconcat = [] as any[];
+  newSpec.vconcat.push({
+    layer: newSpec.layer,
+  });
+  delete newSpec.layer;
```

Comment:
Why are we deleting the `newSpec.layer`?

Reply:
It is so we can nest the `layer` in the `vconcat` field on line 285.